File: [001-149]/[001] Two Sum.cs
/*
Last Solved: 25.10.24
Topics Utilised: Hash Table
Link: https://leetcode.com/problems/two-sum/
Time Complexity: O(n)
Space Complexity: O(n)
*/

public class Solution {
    public int[] TwoSum(int[] nums, int target) {
        //Number, Index
        Dictionary<int, int> dict = new Dictionary<int, int>();
        
        for (int i = 0; i < nums.Length; i++)
        {
            //Number needed
            int diff = target - nums[i];

            //If number was previously seen
            if (dict.ContainsKey(diff)){
                //Return Index of needed num and current index
                return new int[]{dict[diff], i};
            }
            //Number not yet seen
            else{
                //Add current num and index
                dict.Add(nums[i], i);
            }
        }

        return null;
    }
}

File: [001-199]/[001] Two Sum.py
"""
Topics Utilised: Hash Table
Time Complexity: O(n)
Space Complexity: O(n)
"""

class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        
        numsDict = {}
        numsLen = len(nums)
        # for each num in nums
        for i in range(numsLen):
            remainder = target - nums[i]

            # if dict contains target - num
            if remainder in numsDict:
                # return curIndex, dict[num]
                return [i, numsDict[remainder]]
            # else remainder not in dict, add this num to the dict
            else:
                numsDict[nums[i]] = i
